modules/core/src/main/java/org/apache/ignite/internal/util/IgniteSpinReadWriteLock.java

```diff
@@ -0,0 +1,406 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.util;
+
+import java.util.concurrent.TimeUnit;
+import org.apache.ignite.internal.tostring.S;
+
+/**
+ * Spin read-write lock.
+ */
+public class IgniteSpinReadWriteLock {
+    /** */
+    private static final long PENDING_WLOCKS_OFFS;
+
+    /** */
+    private static final long STATE_OFFS;
+
+    /**
+     *
+     */
+    static {
+        try {
+            STATE_OFFS = GridUnsafe.objectFieldOffset(IgniteSpinReadWriteLock.class.getDeclaredField("state"));
+
+            PENDING_WLOCKS_OFFS =
+                GridUnsafe.objectFieldOffset(IgniteSpinReadWriteLock.class.getDeclaredField("pendingWLocks"));
+        }
+        catch (NoSuchFieldException e) {
+            throw new Error(e);
+        }
```

Comment:
AFAIK, we thought to replace Unsafe usage with VarHandle where it is possible.

Reply:
Thank you for pointing this out, I prefer to create ticket and add TODO https://issues.apache.org/jira/browse/IGNITE-15536